### ISSUE TITLE
Ability to turn off on-save compilation

### ISSUE THREAD

**Issue by mikkka** [scalameta/metals-feature-requests#278]
### Is your feature request related to a problem? Please describe.

if I have big project and do refactoring, for example type moving this ends in infinite compilation turns in metals, and takes very long time. It would be great if I could do all my changes, save all files, and then request compilation. 

### Describe the solution you'd like

Command or option to turn off and on auto-compilation.

### Describe alternatives you've considered

go to IntelliJ maybe, it can compile on demand through SBT

### Search terms

autosave, disable, compile, source, scan

**Comment by tgodzik** [comment]
Thanks for reporting! The compilation only occurs after you save the files, so an alternative would be to keep the changes unsaved until you are sure you want to compile them.

Getting a usability of this feature request might be a bit tricky.

**Comment by laughedelic** [comment]
Would it be possible to add a command to "Disconnect from the build server" and some indicator of wher Metals is connected to a build server or not? Then the user can reconnect using an existing command.

**Comment by armanbilge** [comment]
Yes, apologies if it already exists, but I would appreciate a way to "turn off" metals sometimes.

**Comment by sjorn3** [comment]
Hey I'd appreciate this too and would like to add a motivating example.

Our codebase takes 25 minutes to compile from scratch. That's something we're working on but it probably won't get better overnight or anything and I don't think it will ever be super quick.

I have to be very careful whenever doing certain actions to try and prevent it from full recompiling, or I'm out another 25 minutes. In particular if I rebase on master and have a conflict partway through, metals will start recompiling things that would not have needed to be recompiled if the rebase had been successful, and then when the rebase is finished it has to compile it again. If I could temporarily disable the build server for the duration of a rebase, there would be significantly less compile time.

In my particular case I prefer to start the build server manually in a terminal and have metals connect to it. But if I disable that build server metals will just spin up another one. So actually a toggle to simply disable automatic build server generation would be sufficient to fix my issue.

My current solution means fully closing vscode and using some other tool to fix merge conflicts when they are sufficiently large that I start running into this issue.

Metals does have a "disconnect from build server without restarting" option, but this doesn't seem to do anything, possibly a bug?

Thanks,

Sean.

**Comment by vladimir-popov** [comment]
Any updates? I had to switch back to IDEA because I love using Neovim's autosave mode, but it causes 100% CPU load when paired with Metals. It would be great if there were an option to run compilation manually through Metals instead of it happening automatically.

**Comment by tgodzik** [comment]
Normally, compile happens on save, so if you turn off the autosave option you can get close to what you need.

Do you have an idea what gets recompiled that causes the CPU usage? Maybe we could optimize that too?

**Comment by vladimir-popov** [comment]
Honestly, I don't want to disable autosave because, for me, it represents a bigger shift in my workflow than switching to IDEA. Additionally, I prefer a manual compilation approach when working with Zig, as its LSP server is quite basic, and I like using Vim's compiler feature to collect diagnostics. I'd be happy to adopt the same approach for Scala (manually running compilation, I mean).

Regarding compilation, I'm working with a large monorepo on my work, and making changes to any projects near the root of the project hierarchy results in an extremely heavy compilation process.

**Comment by tgodzik** [comment]
By default we should not recompile downstream projects only the current one. There might be a bug in that logic then :thinking: 

We do try to recompile downstream projects if we know that they have errors which might be fixed the changes.

There is also this issues: https://github.com/scalameta/metals/issues/6782 <- it might help especially in large projects

**Comment by vladimir-popov** [comment]
Thanks for sharing a related issue! But anyway, ability to run compilation manually may give us more control on resource consumption and save battery even in simple projects. Moreover, as was mentioned by @sjorn3 resolving merge conflict is one more reason to pause compilation.

**Comment by tgodzik** [comment]
Anyway, I might be able to work on it soon, but the change should simple enough. We can add a [user setting](https://github.com/scalameta/metals/blob/main/metals/src/main/scala/scala/meta/internal/metals/UserConfiguration.scala#L28) to enable manual compilation on save and then ignore it in [here](https://github.com/scalameta/metals/blob/caf518ede2dd831a30d114f8545acd401ed10953/metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala#L947)

**Comment by vladimir-popov** [comment]
sounds awesome! Thank you!

**Comment by abbas-taqvi** [comment]
I am experiencing an issue which IMHO might be solved by this feature, 

I am using scala for spark and since the codebase has features that jdk17 and jdk23 are more strict with (this should be a common issue with using spark along with airflow, kafka and nifi) , so I am stuck with using jdk11 for compilation however, metals LSP client quits with 0 signal if the jdk version isn't 23 when opening a scala file in nvim. Therefore I launch nvim with jdk23 and manually compile on another tmux window whenever I want with jdk11, I have a config for sbt to use jdk11 for compilation

In this scenario auto-compilations just fail for me repeatedly whenever the file gets saved. Is there a workaround or am I missing anything?

**Comment by tgodzik** [comment]
You should be able to use the metals.javaHome setting to set the JDK target you want to compile for even if metals is running on a newer one

**Comment by abbas-taqvi** [comment]
That works like a charm 🥳
Thanks!

**Comment by nedsociety** [comment]
In our case it's a problem during the project import -- each import triggers multiple source generation tasks in sbt, then each file change from that generation triggers recompilation _during_ the import. That slows down the code generation in a feedback loop by 10x compared to when we just call sbt generateSources. And the compilation triggered during the import is useless since they all fail with partially generated sources (like missing files and defs, since ScalaPB tries to remove every managed files and regenerate them one-by-one), so it's better for us to just disable the automatic compilation during the whole import process.

> Anyway, I might be able to work on it soon, but the change should simple enough. We can add a [user setting](https://github.com/scalameta/metals/blob/main/metals/src/main/scala/scala/meta/internal/metals/UserConfiguration.scala#L28) to enable manual compilation on save and then ignore it in [here](https://github.com/scalameta/metals/blob/caf518ede2dd831a30d114f8545acd401ed10953/metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala#L947)

Sounds awesome, can we have something like "Disable during import" as a user setting here?

**Comment by tgodzik** [comment]
> Sounds awesome, can we have something like "Disable during import" as a user setting here?

This might actually be a good default, since in a lot of cases the compilation might fail during reimporting. Certainly worth a normal issue.
